enclave-runner/src/usercalls/mod.rs

```diff
@@ -346,36 +381,221 @@ impl EnclaveState {
+                },
+                (Ok(_), EnclaveEntry::ExecutableNonMain) => {
+                    continue;
+                },
+                (Ok(ret), EnclaveEntry::Library) => {
+                    return Ok(ret);
+                }
+                (Err(e @ EnclaveAbort::Exit {..}), EnclaveEntry::ExecutableNonMain) |
+                (Err(e @ EnclaveAbort::InvalidUsercall(_)), EnclaveEntry::ExecutableNonMain) => {
+                    let cmd = enclave.kind.as_command().unwrap();
+                    let mut cmddata = cmd.data.lock().unwrap();
+
+                    if cmddata.primary_panic_reason.is_none() {
+                        cmddata.primary_panic_reason = Some(e)
+                    } else {
+                        cmddata.other_reasons.push(e)
+                    }
+                    return Err(EnclaveAbort::Secondary);
+                },
+                (Err(EnclaveAbort::Secondary), EnclaveEntry::ExecutableNonMain) => {},
+                (Err(e), EnclaveEntry::ExecutableNonMain) => {
+                    let cmd = enclave.kind.as_command().unwrap();
+                    let mut cmddata = cmd.data.lock().unwrap();
+                    cmddata.other_reasons.push(e)
+                },
+
+            }
+        }
+        unreachable!();
+    }
+
+    fn setup_and_start_work(num_of_worker_threads: usize,
+                            start_work: Work,
+                            enclave: Arc<EnclaveState>
+    ) -> StdResult<(u64, u64), EnclaveAbort<EnclavePanic>> {
+        let (io_queue_send, io_queue_receive) = mpsc::channel();
+
+        let (work_sender, work_receiver) = crossbeam::crossbeam_channel::unbounded();
+        work_sender.send(start_work).expect("Work sender couldn't send data to receiver");
+
+        let join_handlers = EnclaveState::create_worker_threads(num_of_worker_threads, work_receiver,io_queue_send);
+        // main syscall polling loop
+        let main_result = EnclaveState::syscall_loop(io_queue_receive, enclave.clone(), work_sender);
+
+        for handler in join_handlers {
+            let _ = handler.join();
+        }
+        return main_result;
+    }
+
     pub(crate) fn main_entry(
         main: ErasedTcs,
         threads: Vec<ErasedTcs>,
         usercall_ext: Option<Box<dyn UsercallExtension>>) -> StdResult<(), failure::Error>  {
         let mut event_queues =
             FnvHashMap::with_capacity_and_hasher(threads.len() + 1, Default::default());
         let main = Self::event_queue_add_tcs(&mut event_queues, main);
-
-        let threads = threads
+        let threads: Vec<StoppedTcs> = threads
             .into_iter()
             .map(|thread| Self::event_queue_add_tcs(&mut event_queues, thread))
             .collect();
 
+        let start_queue = crossbeam::queue::SegQueue::new();
+        for tcs in threads {
+            start_queue.push(tcs);
+        }
+        let StoppedTcs {
+            tcs: main_tcs,
+            event_queue,
+        } = main;
```

Comment:
I feel like this destructuring doesn't really add much. Just write `main.event_queue` and `main.tcs` below, where needed. Same inside `library_entry` and `launch_thread`.